### DISCUSSION TITLE
Multipart support for RR client

### DISCUSSION THREAD

**Discussion by luca-bassoricci** [quarkusio/quarkus#36780] [Q&A] [answered]
We are moving from 3.2.6.Final to 3.4.3 and we need to change code about multipart because we are still using legacy RR support which doesn't work with new version.
This is our server part which basically read a generic form and need to forward to `MyService` as-is:
```
public void ServerResource {
  public void forwardMultipart(final MultipartFormDataInput request) throws IOException {
    var outputForm = (convert request to org.jboss.resteasy.plugins.providers.multipart.MultipartFormDataOutput);
    myService.forwardForm(outputForm);
  }
}
```
```
@RegisterRestClient
public interface MyService {
  @Path("")
  @PATCH
  @Consumes(MediaType.MULTIPART_FORM_DATA)
  void forwardForm(org.jboss.resteasy.plugins.providers.multipart.MultipartFormDataOutput request);
}
```

Old code relied on legacy rest-client support (https://stackoverflow.com/a/64801524/2587166) and all worked fine but with new version  I'm unable to build-up a generic multipart form programmatically and send via rest-client.
I tried add @RegisterProvider(MultipartMessageBodyWriter.class) but - as expected - doesn't work because it's a server-side writer.
Which is the correct way to proceed?

Regards
Luca

**Comment by luca-bassoricci** [comment]
Any advice? :)

**Comment by luca-bassoricci** [comment]
up

**Comment by luca-bassoricci** [comment]
I have discovered [QuarkusMultipartForm](https://github.com/quarkusio/quarkus/blob/main/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/multipart/QuarkusMultipartForm.java) and I was able to convert a `org.jboss.resteasy.reactive.server.multipart.MultipartFormDataInput` but I have some concerns about using `QuarkusMultipartForm` because it is undocumented.
Any other way to create a generic multipart form that can be send using rest-client as single parameter?

**Reply by doogle-oss** [reply]
Yes , moving from 3.3.1 to 3.6.1 and without legacy rest client now our code does not send List<File> and a String based Json. Can any update be shared on this.

**Comment by doogle-oss** [comment]
Finally i had got success with using #QuarkusMultipartForm

[multipart-quarkus](https://github.com/debu999/multipart-quarkus) 
[multipart-spring](https://github.com/debu999/multipart-spring)

So the following code snippet in needed.
```
@POST
  @Path("/multipart")
  @Produces(MediaType.APPLICATION_JSON)
  @Consumes(MediaType.MULTIPART_FORM_DATA)
  public Map<String, String> multipart(MultiPartPayloadFormData data) {

    QuarkusMultipartForm qmf = new QuarkusMultipartForm();
    qmf.attribute("jsonPayload", data.getJsonPayload(), "jsonPayload");
    data.getFiles().forEach(f -> {
      qmf.binaryFileUpload("file", f.name(), f.filePath().toString(), f.contentType());
    });

    return springService.springMultipartCall(qmf);
  }
```
Spring Controller details
```
  @PostMapping(path = "/multipart",
      consumes = MULTIPART_FORM_DATA_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
  public Map<String, String> multipart(@RequestParam String jsonPayload,
      @RequestPart MultipartFile[] files) {
    return Map.of("fileName", String.join(": ", "File received is : ", files[0].getName()));
  }
```

**Reply by doogle-oss** [reply]
What is interesting is we neednot pass the media type and its auto infered. Now the question is when we can get some documentation.

**Reply by doogle-oss** [reply]
@Sgitario any comments on this

**Reply by Sgitario** [reply]
QuarkusMultipartForm should be the way to go as far as I know. About the documentation updates +100. This class should be documented in the guide and in the migration guide.

**Reply by luca-bassoricci** [reply]
```
MultipartFormDataInput form;
var f = new QuarkusMultipartForm();
for (var e : form.getValues().entrySet())
{
	for (var fv : e.getValue())
	{
		if(fv.isFileItem())
		{
			final var fi = fv.getFileItem();
			var mt = Objects.toString(fv.getHeaders().getFirst(HttpHeaders.CONTENT_TYPE),
					MediaType.APPLICATION_OCTET_STREAM);
			if(fi.isInMemory())
			{
				f.binaryFileUpload(e.getKey(), fv.getFileName(), Buffer.buffer(IOUtils.toByteArray(fi.getInputStream())), mt);
			}
			else
			{
				f.binaryFileUpload(e.getKey(), fv.getFileName(), fi.getFile().toString(), mt);
			}
		}
		else
		{
			f.attribute(e.getKey(), fv.getValue(), fv.getFileName());
		}
	}
}
```
@Sgitario  is this the right way to convert a server-side `MultipartFormDataInput` to `QuarkusMultipartForm`?

**Reply by doogle-oss** [reply]
New Issue Link [multipart unify with @PartType](https://github.com/quarkusio/quarkus/issues/37765)

**Reply by luca-bassoricci** [reply]
> QuarkusMultipartForm should be the way to go as far as I know. About the documentation updates +100. This class should be documented in the guide and in the migration guide.

@Sgitario if `QuarkusMultipartForm` should be the way to use generic multipart-form can be used to substitute `MultipartFormDataInput` in server side? eg:
```
  @POST
  @Path("/multipart")
  @Produces(MediaType.APPLICATION_JSON)
  @Consumes(MediaType.MULTIPART_FORM_DATA)
  public Map<String, String> multipart(QuarkusMultipartForm multipaftForm) {
    return springService.springMultipartCall(multipaftForm);
  }
```

**Reply by doogle-oss** [reply]
ClientMultipartForm is the new form I am documenting it since QuarkusMultipartForm was in impl package we should not use it rather we should use `ClientMultipartForm`

**Comment by luca-bassoricci** [comment]
> Finally i had got success with using #QuarkusMultipartForm
> 
> [multipart-quarkus](https://github.com/debu999/multipart-quarkus) [multipart-spring](https://github.com/debu999/multipart-spring)
> 

multipart-quarkus project gives to me a 404...can you make it available again?
Thanks 
Luca

**Reply by doogle-oss** [reply]
Extremely sorry @luca-bassoricci it was private repo. Made it public you can access.
```
@POST
  @Path("/multipart")
  @Produces(MediaType.APPLICATION_JSON)
  @Consumes(MediaType.MULTIPART_FORM_DATA)
  public Map<String, String> multipart(MultiPartPayloadFormData data) {

    QuarkusMultipartForm qmf = new QuarkusMultipartForm();
    qmf.attribute("jsonPayload", data.getJsonPayload(), "jsonPayload");
    data.getFiles().forEach(f -> {
      qmf.binaryFileUpload("file", f.name(), f.filePath().toString(), f.contentType());
    });

    return springService.springMultipartCall(qmf);
  }
```
![Screenshot 2023-12-13 at 9 18 14 AM](https://github.com/quarkusio/quarkus/assets/1910577/549da347-dafe-4780-806e-6f94a9e560a7)

**Comment by doogle-oss** [comment, marked as answer]
We now have `ClientMultipartForm` to build instead of `QuarkusMultipartForm` This is targetted for 3.7 and the same has been merged to main branch today including documentation.

**Comment by vinimdev** [comment]
Guys, I am using an application exactly with the settings of the apps that @debu999 posted above, but when I send more than one file, it shows a MalformedMessage error, with just one file, it works normally. Does anyone have a solution?

https://github.com/quarkusio/quarkus/discussions/39751
